### PR TITLE
Env: Bind "core" files to tests environment

### LINKED ISSUE #21164: Env: Test site is not updated with changes made in "core" directory
**To reproduce**

1. Create a `.wp-env.json` that has its `"core"` set to a local path:

   ```json
   { "core": "/path/to/wordpress-develop/build" }
   ```
2. Run `wp-env start`.

3. Make changes to something in `wordpress-develop/build`.

The changes appear in `localhost:8888` (the development site) but **not** `localhost:8889`.

**Expected behavior**

Changes should appear in **both** sites.

**Additional context**

This happens because we copy the local `wordpress` directory into `tests-wordpress` when `wp-env` starts. This is necessary so that e.g. `wp-config.php` and `wp-content/uploads` may differ between the two sites.

We will have to find a different way of having the development and tests sites be different instances while at the same time using the same `wordpress` files.

### PR DESCRIPTION
**Pull request by noisysocks**
Fixes https://github.com/WordPress/gutenberg/issues/21164.

When using a local source for "core" we want to ensure two things:

1. That changes the user makes within the "core" directory are served in both the development and tests environments.
2. That the development and tests environment use separate databases and `wp-content/uploads`.

Only (2) was working because the tests environment served everything from a totally separate copy of WordPress.

The fix is to bind every file **except** `wp-config.php` and `wp-content` from the local "core" directory to the tests environment.

**To test:**

1. Create a `.wp-env.json` that has its `"core"` set to a local path:

   ```json
   { "core": "/path/to/wordpress-develop/build" }
   ```
2. Run `wp-env start`.

3. Make changes to something in `wordpress-develop/build`.

The changes should appear in both `localhost:8888` and `localhost:8889`.

We should also verify that this does not break existing flows, e.g. running `wp-env start` with `"core": null` and `"core: "WordPress/WordPress"`.

### CODE REVIEW

File: packages/env/lib/build-docker-compose-config.js [hunk 1]
```diff
@@ -43,13 +43,52 @@ module.exports = function buildDockerComposeConfig( config ) {
...
 
-	const testsMounts = [
-		`${
-			config.coreSource ? config.coreSource.testsPath : 'tests-wordpress'
-		}:/var/www/html`,
-		...pluginMounts,
-		...themeMounts,
-	];
+	let testsMounts;
+	if ( config.coreSource ) {
```
Comment (noisysocks) [inline packages/env/lib/build-docker-compose-config.js]: Hm, actually, we probably should only do the crazy bindings below when `config.coreSource && config.coreSource.type === 'local'`. Non-local core sources can rely on the copy.

Comment (noisysocks) [reply packages/env/lib/build-docker-compose-config.js]: ...or, we could leave it like this adapt `copyWordPress` to only copy the directory plus `wp-config.php`, `wp-config-sample.php` and `wp-content`. That would mean `wp-env start` is faster which is nice.

File: packages/env/lib/build-docker-compose-config.js [hunk 2]
```diff
@@ -43,13 +43,52 @@ module.exports = function buildDockerComposeConfig( config ) {
...
+			//
+			// - wp-config.php        <- $tests-wordpress/wp-config.php
+			// - wp-config-sample.php <- $tests-wordpress/wp-config.php
+			// - wp-content           <- $tests-wordpress/wp-content
+			// - *                    <- $wordpress/*
+			//
+			// https://github.com/WordPress/gutenberg/issues/21164
+			`${ config.coreSource.testsPath }:/var/www/html`,
+			...fs
+				.readdirSync( config.coreSource.path )
```
Comment (noisysocks) [inline packages/env/lib/build-docker-compose-config.js]: I don't really like how `https://github.com/WordPress/gutenberg/pull/21195` has become an "impure" function.

Comment (noisysocks) [reply packages/env/lib/build-docker-compose-config.js]: Ah, this won't work when the core source isn't local as those files won't have been downloaded yet.

File: packages/env/lib/build-docker-compose-config.js (hunk 1 above)
Comment (noisysocks) [reply packages/env/lib/build-docker-compose-config.js]: Fixed in c97ab0715a.

**Comment by noisysocks** [comment]
OK, I'm happy with this.

> We should also verify that this does not break existing flows, e.g. running wp-env start with "core": null and "core: "WordPress/WordPress".

I tested this and all seems gucci 👌.

**Review by tellthemachines — APPROVED** [review]
I've tested this with my local dev env, and running e2e tests on core now works 🎉 
I also tested a few of the custom configs documented on the env package readme and they built successfully.
Code looks good.

**Pull request merged**